### PR TITLE
update(falco_metrics): rearrange evts and drops prometheus metrics

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

> /kind release

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

/area engine

> /area tests

> /area proposals

> /area CI

**What this PR does / why we need it**:

Rearrange evts and drops prometheus metrics,
Fixes https://github.com/falcosecurity/falco/issues/3309

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
update(falco_metrics)!: rearrange `n_evts_cpu` and `n_drops_cpu` Prometheus metrics to follow best practices
```

### CODE REVIEW

File: userspace/engine/falco_utils.cpp [hunk 1]
```diff
@@ -151,7 +151,7 @@ std::string calculate_file_sha256sum(const std::string& filename)
 }
 #endif
 
-std::string sanitize_metric_name(const std::string& name)
+std::string sanitize_rule_name(const std::string& name)
```
Comment (incertum) [inline userspace/engine/falco_utils.cpp:154]: We still need this for the JSON output metrics even though we moved the Prometheus sanitization to libs. I renamed it to better reflect for what it is used.

**Comment by leogr** [comment]
> This PR may bring feature or behavior changes in the Falco engine and may require the engine version to be bumped.
> 
> Please double check **userspace/engine/falco_engine_version.h** file. See [versioning for FALCO_ENGINE_VERSION](https://github.com/falcosecurity/falco/blob/master/RELEASE.md#falco-repo-this-repo).
> 
> /hold

false positive
/unhold

File: userspace/falco/falco_metrics.cpp [hunk 2]
```diff
@@ -246,21 +248,68 @@ std::string falco_metrics::to_text(const falco::app::state& state)
...
+			if (strncmp(metric.name, "n_evts_cpu", 10) == 0 || strncmp(metric.name, "n_drops_cpu", 11) == 0) // prefix match
+			{
+				std::string name_str(metric.name);
+				re2::RE2 pattern("(\\d+)");
+				std::string cpu_number;
+				if (re2::RE2::PartialMatch(name_str, pattern, &cpu_number))
+				{
+					re2::RE2::GlobalReplace(&name_str, pattern, "");
+					auto metric_new = libs::metrics::libsinsp_metrics::new_metric(name_str.c_str(),
+								METRICS_V2_KERNEL_COUNTERS, // todo replace with new METRICS_V2_KERNEL_COUNTERS_PER_CPU after bumping libs the next time
```
Comment (FedeDP) [inline userspace/falco/falco_metrics.cpp:272]: We bumped libs in Falco master to 0.18.0-rc1 ;)

File: userspace/falco/falco_metrics.cpp [hunk 3]
```diff
@@ -246,21 +248,68 @@ std::string falco_metrics::to_text(const falco::app::state& state)
...
 			}
 
 			if (metric.flags & METRICS_V2_PLUGINS)
 			{
-				namespace_name = "plugins";
+				prometheus_subsystem = "plugins";
 			}
 
-			prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metric, "falcosecurity", namespace_name);
+			if (strncmp(metric.name, "n_evts_cpu", 10) == 0 || strncmp(metric.name, "n_drops_cpu", 11) == 0) // prefix match
```
Comment (FedeDP) [inline userspace/falco/falco_metrics.cpp:263]: Here a comment with a small example of `metric.name` would better clarify this code!

File: userspace/falco/falco_metrics.cpp [hunk 4]
```diff
@@ -246,21 +248,68 @@ std::string falco_metrics::to_text(const falco::app::state& state)
...
+								metric.value.u64);
+					const std::map<std::string, std::string>& const_labels = {
+						{"cpu", cpu_number}
+					};
+					prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metric_new, "falcosecurity", prometheus_subsystem, const_labels);
+				}
+			}
+			else if (strncmp(metric.name, "n_drops_buffer_total", 21) == 0) // exact match
+			{
+				continue;
```
Comment (FedeDP) [inline userspace/falco/falco_metrics.cpp:285]: A small comment about why we are skipping this?

File: userspace/falco/falco_metrics.cpp [hunk 5]
```diff
@@ -246,21 +248,68 @@ std::string falco_metrics::to_text(const falco::app::state& state)
...
+								METRIC_VALUE_UNIT_COUNT,
+								METRIC_VALUE_METRIC_TYPE_MONOTONIC,
+								metric.value.u64);
+					const std::map<std::string, std::string>& const_labels = {
+						{"cpu", cpu_number}
+					};
+					prometheus_text += prometheus_metrics_converter.convert_metric_to_text_prometheus(metric_new, "falcosecurity", prometheus_subsystem, const_labels);
+				}
+			}
+			else if (strncmp(metric.name, "n_drops_buffer_total", 21) == 0) // exact match
```
Comment (FedeDP) [inline userspace/falco/falco_metrics.cpp:283]:
```suggestion
			else if (strcmp(metric.name, "n_drops_buffer_total") == 0) // exact match
```
Using `strcmp` clarifies that we are looking for an exact match and not a prefix match imo!

File: falco.yaml [hunk 6]
```diff
@@ -1104,6 +1107,8 @@ metrics:
   resource_utilization_enabled: true
   state_counters_enabled: true
   kernel_event_counters_enabled: true
+  # Enabling `kernel_event_counters_per_cpu_enabled` automatically enables `kernel_event_counters_enabled`
+  kernel_event_counters_per_cpu_enabled: false
```
Comment (incertum) [inline falco.yaml:1111]: CC @Andreagit97

**Review by leogr — DISMISSED** [review]

**Review by FedeDP — DISMISSED** [review]
/approve

File: userspace/falco/configuration.cpp
[the remark this answers is not in the document]
Comment (FedeDP) [reply userspace/falco/configuration.cpp:1]: /hold

[the remark this answers is not in the document]
Comment (FedeDP) [reply userspace/falco/configuration.cpp:1]: Updated :)
/unhold

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**